Pass readable product id to checkout page in URL

umarmughal824 opened the pull request:
#### Pre-Flight checklist

- [ ] Screenshots and design review for any changes that affect layout or styling
  - [ ] Desktop screenshots
  - [ ] Mobile width screenshots
  - [ ] Tag @ferdi or @pdpinch for review
- [ ] Migrations
  - [ ] Migration is backwards-compatible with current production code
- [ ] Testing
  - [ ] Code is tested
  - [ ] Changes have been manually tested
- [ ] Settings
  - [ ] New settings are documented and present in `app.json`
  - [ ] New settings have reasonable development defaults, if applicable

#### What are the relevant tickets?
https://trello.com/c/8pmf7Bnk/32-pass-readable-product-id-to-checkout-page-in-url

#### What's this PR do?
Pass readable product id to checkout page in URL

#### How should this be manually tested?
Try to enroll in a course, now it should show readable_id rather than product_id in the link.
Or open the course with `http://xpro.odl.local:8053/checkout/?product={readable_id}`.

asadiqbal08 commented on ecommerce/serializers.py:
```diff
@@ -365,17 +365,22 @@ def _update_items(cls, basket, items):
+            product = models.Product.objects.filter(
+                dj_models.Q(courseruns__courseware_id=readable_id) |
+                dj_models.Q(programs__readable_id=readable_id)
+            ).first()
```
@umarmughal824 be sure when user will click on `Enroll Now` button then either there should be the support of `Product_id` too or construct the URL with `readable_id` from that page.
Other wise I am sure, 'No item in basket' will be shown over the checkout page.

umarmughal824 replied:
yes replacing that part with `readable_id`.

asadiqbal08 commented:
Instead of showing me `Invalid product id` it is rendering the checkout page with the item. 
`/checkout/?product_id=412312323232` 
###### NOTE First load the basket with proper valid id then change it in the URL.

umarmughal824 commented:
> @umarmughal824
> 
> * First, I would like to request you please add the support to `URL` with `checkout/?product=4` instead of `checkout/?product_id=4` because the url constructed behind the `Enroll Now` button following the `checkout/?product=4` approach.
> * and I would prefer, If we also construct the url over the Enroll Now button with readable format e.g. `course-v1:TestX+SysEngxB2+1T2019` instead of using integer id there.

It's working fine on my system after reloading the server.

umarmughal824 commented:
> Instead of showing me `Invalid product id` it is rendering the checkout page with the item.
> `/checkout/?product_id=412312323232`
> 
> ###### NOTE First load the basket with proper valid id then change it in the URL with random string.

That case was not in my mind or was not there on the ticket, so it not handled as of now. Now going to address it.

umarmughal824 commented:
> Instead of showing me `Invalid product id` it is rendering the checkout page with the item.
> `/checkout/?product_id=412312323232`
> 
> ###### NOTE First load the basket with proper valid id then change it in the URL with random string.

I have debugged the code and verified in master, it's not an issue so its the identical behavior of master branch. So, I am assigning it back to `Needs Review`.

umarmughal824 commented:
> I just want to check that this doesn't break existing URLs that may have been distributed over email already, e.g.
> 
> https://xpro.mit.edu/checkout/?product=52&amp;code=3z7524a20f024d21ab8363885907864b

Currently I have removed the support of `product=product_id` and replaced it with `product=readable_id`. Will add the support of product_id back. Now what about the Enroll, checkout button etc what do we need to show in them either product_id or readable_id?

umarmughal824 commented:
> > I just want to check that this doesn't break existing URLs that may have been distributed over email already, e.g.
> > https://xpro.mit.edu/checkout/?product=52&amp;code=3z7524a20f024d21ab8363885907864b
> 
> @pdpinch Currently I have removed the support of `product=product_id` and replaced it with `product=readable_id`. Will add the support of product_id back. Now what about the Enroll, checkout button etc what do we need to show in them either `product_id` or `readable_id`?
> 
> Or do we want `readable_id` only for the checkout page as a query parameter and keep all the enrollment button links with `product_id`?

@pdpinch I need your opinion here.

pdpinch commented:
The goal here was to support a more human-readable URL for the checkout URL. It’s not necessary to change to behavior of the enrollment button or the enrollment email, which work fine. This is just a convenience for people who need to write checkout URLs like customer support.
